### PR TITLE
_memoria_matchups algorithm update

### PR DESCRIPTION
**Pull request by MemoriAeni**
- over exagerated win prediction the higher it is for 2v2 and 3v3, this allows the win prediction to be higher than for each individual on a team if they're a premade
- matchup weight now takes into consideration the maximum rating in each pair as well as the rating difference
- adjusted some variables until it somewhat worked (lol), overall smurfs and solo players should gain a quite bit less rating
- maximum rating change decreases above 5k (1/x where x = rating in thousands above 5k -> x0.5 multiplier at 9k rating) and under 1.5k (1/x where x = rating in thousands below 1.5k -> ~x0.625 multiplier at 1 rating)

### CODE REVIEW

File: calculators/_memoria_matchups.php [hunk 1]
```diff
@@ -20,38 +21,45 @@ function _memoria_matchups__main(&$teams) {
...
     $matchup_predictions_weight = [];
     $matchup_predictions_weight_total = 0;
     for ($i = 0; $i < count($teams[0]); $i++) {
         $elo_part = abs(($teams[0][$i]['rating'] - $teams[1][$i]['rating']) / $factor);
         $matchup_predictions[$i] = floor(1000 / (1 + 10 ** $elo_part)) / 1000;
-		
+        if (count($teams[0]) > 1) {
+            $matchup_predictions[$i] -= $matchup_predictions[$i] ** (2 / (0.05 + $elo_part));
+            if ($matchup_predictions[$i] < 0) {
+                $matchup_predictions[$i] = 0;
```
Comment (dima-stefantsov) [inline calculators/_memoria_matchups.php]: $matchup_predictions[$i] = max(0, $matchup_predictions[$i]);

File: calculators/_memoria_matchups.php [hunk 2]
```diff
@@ -20,38 +21,45 @@ function _memoria_matchups__main(&$teams) {
...
+            }
+        }
+
         if ($teams[0][$i]['rating'] > $teams[1][$i]['rating']) {
             $matchup_predictions[$i] = 1 - $matchup_predictions[$i];
         }
-        
-        $matchup_predictions_weight[$i] = ((abs(0.5 - $matchup_predictions[$i]) + 0.015) * 6) ** ($elo_part * 1.5);
+
+        $matchup_predictions_weight[$i] = ((abs(0.5 - $matchup_predictions[$i]) + 1) * 3) ** ($elo_part / 1.25 + (max($teams[0][$i]['rating'], $teams[1][$i]['rating']) / $factor));
```
Comment (dima-stefantsov) [inline calculators/_memoria_matchups.php]: `(max($teams[0][$i]['rating'], $teams[1][$i]['rating']) / $factor)` extra parentheses are fine for readability, but +- and */ operations priority are very clear, do not require extra parentheses.

To visually guide the reader, you may omit the spaces around / here, `** ($elo_part/1.25 + max($teams[0][$i]['rating'], $teams[1][$i]['rating'])/$factor)`

File: calculators/_memoria_matchups.php [hunk 3]
```diff
@@ -20,38 +21,45 @@ function _memoria_matchups__main(&$teams) {
...
-    
+
     $prediction = 0;
     for ($i = 0; $i < count($teams[0]); $i++) {
         $normalised_weight_mult = $matchup_predictions_weight[$i] / $matchup_predictions_weight_total;
         $prediction += $matchup_predictions[$i] * $normalised_weight_mult;
     }
-    
+
+    $max_reward = 250 / max(1, max(1.25, from($teams[0])->max('$v["rating"]') / $factor, from($teams[1])->max('$v["rating"]') / $factor) - 0.25, max(1, $factor / (from($teams[0])->min('$v["rating"]') + 2500), $factor / (from($teams[1])->min('$v["rating"]') + 2500)));
```
Comment (dima-stefantsov) [inline calculators/_memoria_matchups.php]: I'll accept it if you want, but I advise you separating this long formula into like 5 separate lines, to make it easier to read and understand.

File: calculators/_memoria_matchups.php (hunk 1 above)
Comment (MemoriAeni) [reply calculators/_memoria_matchups.php]: changed

File: calculators/_memoria_matchups.php (hunk 2 above)
Comment (MemoriAeni) [reply calculators/_memoria_matchups.php]: I think I experimented with another variable in there and forgot to remove the parentheses..
either way, changed

File: calculators/_memoria_matchups.php (hunk 3 above)
Comment (MemoriAeni) [reply calculators/_memoria_matchups.php]: Separated it into three
```
    $max_reward = 250 / max(1, 
    max(1.25, from($teams[0])->max('$v["rating"]') / $factor, from($teams[1])->max('$v["rating"]') / $factor) - 0.25,
    max(1, $factor / (from($teams[0])->min('$v["rating"]') + 2500), $factor / (from($teams[1])->min('$v["rating"]') + 2500)));

Comment (dima-stefantsov) [reply calculators/_memoria_matchups.php]: I mean spreading it into separate lines with assigning different variables. Like
```
$min1 = ;
$min2 = ;
$team1_coefficient = ;
$team2_coefficient = ;
$max_reward = $min1/$min2 ** ($team1_c * $team2_c)
```

if you want to do a single formula multiline like you did, following lines should be shifted right 4 spaces.

Comment (MemoriAeni) [reply calculators/_memoria_matchups.php]: Ah, misunderstood.
Will keep it in a single formula, shifted the 2 lines

**Pull request merged**